### PR TITLE
Fix VB nameof inside async or iterator

### LINKED ISSUE #23019: EV Error BC36946 using NameOf with Async Function
**Version Used**: 
VB 15.4
**Steps to Reproduce**:
1. copy program fragment below in VB code window
```
       Private Async Function B_Async(_document As Document, _CancellationToken As CancellationToken) As Task(Of Document)
            Debug.WriteLine($"#Codefix Entering {NameOf(B_Async)}")
            Return Nothing
        End Function
```
**Expected Behavior**:
No Errors
**Actual Behavior**:
Error BC36946 The implicit return variable of an Iterator or Async method cannot be accessed.

### PR DESCRIPTION
**Pull request by hartmair**
Suppress BC 36946 in case of nameof argument, fixes #23019

### CODE REVIEW

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/NameOfTests.vb [hunk 1]
```diff
@@ -3637,5 +3637,32 @@ BC30469: Reference to a non-shared member requires an object reference.
...
+    End Sub
+    Async Function M() As Task
+        Console.WriteLine(NameOf(M))
+        Await Task.Delay(0)
+    End Function
+End Module
+    </file>
+</compilation>
+
+            Dim comp = CreateCompilationWithMscorlib40AndVBRuntimeAndReferences(compilationDef, {MscorlibRef_v4_0_30316_17626, MsvbRef_v4_0_30319_17929}, TestOptions.DebugExe)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/NameOfTests.vb:3661]: >CreateCompilationWithMscorlib40AndVBRuntimeAndReferences [](start = 23, length = 56)

Please use helper that is not tied to any specific mscorlib version

Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/NameOfTests.vb:3661]: >{MscorlibRef_v4_0_30316_17626, MsvbRef_v4_0_30319_17929} [](start = 96, length = 56)

This looks suspicious. Why is this needed?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

**Comment by AlekseyTs** [comment]
@hartmair Would it be fine if we "squash" commits in this PR during merging?

**Comment by hartmair** [comment]
@AlekseyTs I would also prefer a squash commit

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/NameOfTests.vb (hunk 1 above)
Comment (hartmair) [reply src/Compilers/VisualBasic/Test/Semantic/Semantics/NameOfTests.vb]: I had copy/pasted this from https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Test/Semantic/Semantics/AsyncAwait.vb tests.

**Comment by hartmair** [comment]
@AlekseyTs I would like to mention that almost all existing tests use this `CreateCompilationWithMscorlib40AndVBRuntime` and friends helpers, see e.g. [TypeOfTests.vb](https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Test/Semantic/Semantics/TypeOfTests.vb). The same holds even in the C# tests, see e.g. [NameOfTests.cs](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs). Should I file an issue to refactor these as well to not use hard-coded mscorlib versions?

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/NameOfTests.vb (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Test/Semantic/Semantics/NameOfTests.vb]: > I had copy/pasted this from [https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Test/Semantic/Semantics/AsyncAwait.vb](https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Test/Semantic/Semantics/AsyncAwait.vb) tests.

I can believe this, at the time when the tests were created, that probably was the way to go. It is no longer necessary to do now.

---
In reply to: [451269169](https://github.com/dotnet/roslyn/pull/45738#discussion_r451269169) [](ancestors = 451269169)

**Comment by AlekseyTs** [comment]
> Should I file an issue to refactor these as well to not use hard-coded mscorlib versions?

I don't think this is necessary. The way the old tests are written isn't necessary wrong, just not "modern". I also think that, as we speak @japarson is working on refactoring what references our tests are using. It is quite possible that the test are going to be naturally "modernized" in the process.

**Comment by AlekseyTs** [comment]
@hartmair Thanks for the contribution!

**Comment by japarson** [comment]
@AlekseyTs I think you meant to notify @jaredpar instead of me. 😜

**Comment by AlekseyTs** [comment]
Yes, sorry for the mistake.

**Pull request merged**